internal/document/block.go

```diff
@@ -205,67 +249,210 @@ func (b *CodeBlock) Name() string {
-	if n.Type() == ast.TypeBlock {
-		for i := 0; i < n.Lines().Len(); i++ {
-			line := n.Lines().At(i)
-			_, _ = w.Write(line.Value(source))
-		}
+func newMarkdownBlock(source []byte, node ast.Node) *MarkdownBlock {
+	return &MarkdownBlock{
+		inner:  node,
+		source: source,
+
+		start: -1,
+		stop:  -1,
 	}
+}
 
-	for c := n.FirstChild(); c != nil; c = c.NextSibling() {
-		writeLines(w, c, source)
+func findFirstBlock(node ast.Node) ast.Node {
+	if !node.HasChildren() || node.FirstChild().Type() != ast.TypeBlock {
+		return node
+	}
+	node = node.FirstChild()
+	for node.FirstChild() != nil && node.FirstChild().Type() == ast.TypeBlock {
+		node = node.FirstChild()
 	}
+	return node
+}
+
+func findLastBlock(node ast.Node) ast.Node {
+	if !node.HasChildren() || node.LastChild().Type() != ast.TypeBlock {
+		return node
+	}
+	node = node.LastChild()
+	for node.LastChild() != nil && node.LastChild().Type() == ast.TypeBlock {
+		node = node.LastChild()
+	}
+	return node
+}
+
+func peekLine(b []byte) []byte {
+	idx := bytes.IndexByte(b, '\n')
+	if idx == -1 {
+		return b
+	}
+	return b[:idx]
 }
 
 func (b *MarkdownBlock) Content() string {
 	if b.content == "" {
-		var content strings.Builder
-		writeLines(&content, b.inner, b.source)
-		b.content = content.String()
+		switch b.inner.Kind() {
+		case ast.KindHeading:
+			start := b.inner.Lines().At(0).Start
+			if start > 0 {
+				if idx := bytes.LastIndexByte(b.source[0:start], '\n'); idx != -1 {
+					start = idx + 1
+				} else {
+					start = 0
+				}
+			}
+
+			stop := b.inner.Lines().At(b.inner.Lines().Len() - 1).Stop
+			if idx := bytes.IndexByte(b.source[stop:], '\n'); idx != -1 {
+				stop = stop + idx
+			}
+
+			if line := peekLine(b.source[stop+1:]); bytes.ContainsAny(line, "=-") {
+				stop += len(line) + 1
+			}
+
+			b.content = string(b.source[start:stop])
+			b.start = start
+			b.stop = stop
+		case ast.KindList, ast.KindBlockquote, ast.KindParagraph:
+			firstChild := findFirstBlock(b.inner)
+			lastChild := findLastBlock(b.inner)
+
+			start := firstChild.Lines().At(0).Start
+			if start > 0 {
+				if idx := bytes.LastIndexByte(b.source[0:start], '\n'); idx != -1 {
+					start = idx + 1
+				} else {
+					start = 0
+				}
+			}
+
+			stop := lastChild.Lines().At(lastChild.Lines().Len() - 1).Stop
+			if idx := bytes.IndexByte(b.source[stop:], '\n'); idx != -1 {
+				stop = stop + idx
+			}
+
+			b.content = string(b.source[start:stop])
+			b.start = start
+			b.stop = stop
+		case ast.KindThematicBreak:
+			previousBlock := b.inner.PreviousSibling()
+			for previousBlock != nil && !hasLine(previousBlock) {
+				previousBlock.PreviousSibling()
+			}
+
+			start := 0
+			if previousBlock != nil {
+				_, start = getContentRange(b.source, previousBlock)
+				start++
+			}
+			for ch := b.source[start]; ch == '\n'; ch = b.source[start] {
+				start++
+			}
+
+			nextBlock := b.inner.NextSibling()
+			for nextBlock != nil && !hasLine(nextBlock) {
+				nextBlock.NextSibling()
```

Comment:
```suggestion
				nextBlock = nextBlock.NextSibling()
```
i believe this will fix it